### PR TITLE
Mutable mfds

### PR DESCRIPTION
**Pull request by angri**
This makes Gutenberg-Richter MFD objects easy to manipulate in terms of applying uncertainties. It is needed to simplify logic trees processing.

### CODE REVIEW

**Comment by larsbutler** [comment]
What about tests?

**Comment by angri** [comment]
Doesn't seem like that part of opensha is tested. Honestly I don't
feel like starting to test opensha from scratch...

2011/10/5 Lars Butler
reply@reply.github.com:

**Comment by larsbutler** [comment]
You are correct, many parts of OpenSHA are untested or under-tested. However, do you think that 'lack of sufficient existing test coverage' is a valid reason for _not_ improving the test coverage? I don't.

At any rate, this is not trivial code. Without tests we have no way of knowing if the formulae implemented here are correct, nor do we know for sure if the side effects are correct.

File: java/org/opensha/sha/magdist/GutenbergRichterMagFreqDist.java [hunk 1]
```diff
@@ -141,6 +141,85 @@ public class GutenbergRichterMagFreqDist extends IncrementalMagFreqDist {
...
+
+        // changing the bins array
+        set(magLower, newMagUpper, numVal);
+
+        // preserving the same total moment rate
+        setAllButTotCumRate(magLower, newMagUpper, totMoRate, bValue);
+    }
+
+    /**
+     * Replace values of a and b with provided numbers keeping the total moment
```
Comment (monellid) [inline java/org/opensha/sha/magdist/GutenbergRichterMagFreqDist.java]: The comment is wrong. In the algorithm the total moment rate is not conserved. You compute a new one with the new values of a and b.

File: java/org/opensha/sha/magdist/GutenbergRichterMagFreqDist.java [hunk 2]
```diff
@@ -141,6 +141,85 @@ public class GutenbergRichterMagFreqDist extends IncrementalMagFreqDist {
...
+        // round new magUpper with respect to delta
+        newMagUpper = Math.round(newMagUpper / delta) * delta;
+        // move back to bin center
+        newMagUpper = newMagUpper - delta / 2;
+
+        // setting new computed value
+        setMagUpper(newMagUpper);
+    }
+
+    /**
```
Comment (monellid) [inline java/org/opensha/sha/magdist/GutenbergRichterMagFreqDist.java:168]: Absolute uncertainties do not need to conserve total moment rate for the moment.
Before calculating the number of magnitude values you should probably also round (with respect to delta) the newMagUpper value. Or do you assume that the newMagUpper is rounded before?
If you want to update the maximum magnitude without conserving the total moment rate, what you can do is to compute the a value using the following formula:
a = Log10( (TMR \* (1.5 - b)) / (10^(1.5 - b)_Mmax - 10^(1.5 - b)_Mmin) ) - 16.1 - Log10(b)
where TMR is the total moment rate from the current mfd object
Then compute the total cumulative rate:
TotCumRate = 10^(a -b *Mmin) - 10^(a -b \* Mmax)
where Mmin is the minimum magnitude and Mmax is the NEW maximum magnitude (newMagUpper)
then you can call the method:
setAllButTotMoRate(double magLower, double magUpper,
            double totCumRate, double bValue) 
where you can pass the new magUpper and the new totCumRate

Comment (monellid) [inline java/org/opensha/sha/magdist/GutenbergRichterMagFreqDist.java:168]: Sorry the correct formula for a value is as follows:
a = Log10( (TMR \* (1.5 - b)) / (10^[(1.5 - b)_Mmax] - 10^[(1.5 - b)_Mmin)] ) - 9.05 - Log10(b)

**Comment by monellid** [comment]
The science is ok

File: java_tests/org/opensha/sha/magdist/GutenbergRichterMagFreqDistTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+import org.opensha.sha.magdist.GutenbergRichterMagFreqDist;
+
+
+public class GutenbergRichterMagFreqDistTest {
+
+    @Test
+    public void CheckIncrementMagUpper() {
```
Comment (larsbutler) [inline java_tests/org/opensha/sha/magdist/GutenbergRichterMagFreqDistTest.java]: Nitpick: The first character of the method name should not be capitalized (convention). It should be: `checkIncrementMagUpper()`

**Pull request merged**